### PR TITLE
DateTimePicker: Alternate timezones now behave correctly

### LINKED ISSUE #86305: DateTimePicker: Doesn't update correctly when custom `timeZoneResolver` is set
steps to reproduce in storybook:
- add this to `.storybook/preview.ts`:
```ts
import { setTimeZoneResolver } from '@grafana/data';

setTimeZoneResolver(() => 'America/Indiana/Marengo');
```
- set today in `DateTimePicker.story.tsx` to `const today = '2024-04-16T08:44:41.000000Z';`
- modify the basic story in `DateTimePicker.story.tsx` to:
```tsx
export const Basic: StoryFn<typeof DateTimePicker> = ({ label, minDate, maxDate, showSeconds }) => {
  const [date, setDate] = useState<DateTime>(dateTime(today));
  // the minDate arg can change from Date object to number, we need to handle this
  // scenario to avoid a crash in the component's story.
  // const minDateVal = typeof minDate === 'number' ? new Date(minDate) : minDate;
  // const maxDateVal = typeof maxDate === 'number' ? new Date(maxDate) : maxDate;

  return (
    <DateTimePicker
      label={label}
      // minDate={minDateVal}
      // maxDate={maxDateVal}
      date={date}
      showSeconds={showSeconds}
      onChange={(newValue) => {
        action('on change')(newValue);
        setDate(newValue);
      }}
    />
  );
};
```
- run `yarn storybook` and navigate to the `DateTimePicker` basic story
- click into the input then blur it
- expected: time shouldn't change
- actual: time changes

looks like the timezone information is getting lost in the `onChange` loop

### PR DESCRIPTION
**Pull request by w1kman**
**What is this feature?**
It's a bugfix for the timezone [issue](https://github.com/grafana/grafana/issues/86305) with the `DateTimePicker`

**Why do we need this feature?**
Currently, users that have one timezone in Grafana (`bootData`) and another on in their client/browser, cannot use the `DateTimePicker` as it changes the date `onBlur` as well as showing invalid date/time information in date/time modal.

**Who is this feature for?**
Developers who wants to use `DateTimePicker` as well as users that already are interacting with `DateTimePicker` in published plugins

**Which issue(s) does this PR fix?**:
[86305](https://github.com/grafana/grafana/issues/86305)

Fixes https://github.com/grafana/grafana/issues/86305

**Special notes for your reviewer:**
`setTimeZoneResolver` needs to be used to change timezone.
There are notes in #86305 how you could do this.

I just added the blow snippe to `DateTimePicker.story.tsx`:
```
  useEffect(() => {
    setTimeZoneResolver(() => 'America/Indiana/Marengo')
    return () => setTimeZoneResolver(() => 'browser')
  }, []);
```

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by w1kman — COMMENTED** [review]
self review ✅

**Comment by w1kman** [comment]
Noticed some issues with invalid dates.
Working on a fix.

**Comment by w1kman** [comment]
> Noticed some issues with invalid dates.
Working on a fix.

A fix has been pushed.

**Review by ashharrison90 — COMMENTED** [review]
ok i mostly love this (especially the tests! 🙌) and it does fix the problem, but not in love with the string splitting and chain of useEffects 😅 

have added some more test cases that fully capture the bug. i'm going to play around a little bit more and see if we can refactor into something that behaves **and** looks nice 🙏

File: packages/grafana-ui/src/components/DateTimePickers/DateTimePicker/DateTimePicker.tsx [hunk 1]
```diff
@@ -276,43 +277,68 @@ const DateTimeCalendar = React.forwardRef<HTMLDivElement, DateTimeCalendarProps>
...
-      }
-    }, []);
+    useEffect(() => {
+      setDateValue(timeValue);
+    }, [timeValue]);
 
-    const onChangeTime = useCallback((date: DateTime) => {
-      setInternalDate(date.toDate());
-    }, []);
+    useEffect(() => {}, []);
```
Comment (w1kman) [inline packages/grafana-ui/src/components/DateTimePickers/DateTimePicker/DateTimePicker.tsx:296]: Did I add this? 😅

**Comment by ashharrison90** [comment]
@w1kman ok i've got some proposed changes here: https://github.com/grafana/grafana/pull/86854

this basically gets us down to the minimal amount of changes i think we need

the diff between main ends up being very small: https://github.com/grafana/grafana/compare/main...ash/proposed-changes

**Review by ashharrison90 — APPROVED** [review]
looks great 🙌 thanks for taking the initiative and helping iterate on this @w1kman 🙇

**Pull request merged**